### PR TITLE
VIH-6931 - mapper refactor with logging

### PR DESCRIPTION
**Pull request by levibutler**
### JIRA link (if applicable) ###
VIH-6931

### Change description ###
Our normal logging around the controllers

Mapper changes
I have changed the mappers so they need to be injected
`IMapTo<TIn, TOut>` (up to 5 TIns supported)
or they can be resolved using the `IMapperFactory`

The registrations of the mappers and decorators are handled just implements the new interface
All mappers are decorated with `MapperLoggingDecorator<TIn, TOut> : IMapTo<TIn, TOut>`
This logs before and after mapping. We can later expand on this to globally add additional logging details or exception handling etc.
Future **unit tests** that have a dependency on a mapper should have the mapper response mocked. The mapper should be tested separately (as it currently is)
Now mappers aren't static, we can inject dependencies if needed (like other mappers)

AutoMock
Ive also added automock where i had to change the tests due to ctor changes. This will make the tests less brittle, ive been slowly introducing.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: VideoWeb/VideoWeb/Controllers/ConferencesController.cs [hunk 1]
```diff
@@ -27,12 +29,30 @@ public class ConferencesController : Controller
...
+        private readonly IMapTo<ConferenceForVhOfficerResponse, ConferenceForAdminResponse> _conferenceForVhOfficerResponseMapper;
+        private readonly IMapTo<ConferenceResponseVho, ConferenceDetailsResponse> _conferenceResponseVhoMapper;
+        private readonly IMapTo<ConferenceResponse, ConferenceDetailsResponse> _conferenceResponseMapper;
 
-        public ConferencesController(IVideoApiClient videoApiClient, ILogger<ConferencesController> logger, IConferenceCache conferenceCache)
+        public ConferencesController(
+            IVideoApiClient videoApiClient,
+            ILogger<ConferencesController> logger,
+            IConferenceCache conferenceCache,
+            IMapTo<ConferenceForJudgeResponse, JudgeConference> conferenceForJudgeResponseMapper,
```
Comment (shaed-parkar) [inline VideoWeb/VideoWeb/Controllers/ConferencesController.cs:42]: Any chance we can avoid listing so many mappers in the ctor but rather a single IMapTo factory

**Review by shaed-parkar — COMMENTED** [review]
Nice work, though a factory would make the controllers a lot leaner

File: VideoWeb/VideoWeb/Controllers/ConferencesController.cs (hunk 1 above)
Comment (levibutler) [reply VideoWeb/VideoWeb/Controllers/ConferencesController.cs]: changed

File: VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public class LoggingDataExtractor : ILoggingDataExtractor
+    {
+        public Dictionary<string, object> ConvertToDictionary(object input, string path = null, int debth = 0)
+        {
+            var result = new Dictionary<string, object>();
+            if (debth > 3)
+            {
+                // Protection from recusrive properties
```
Comment (virensareen) [inline VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs:19]: Comment?

File: VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                {
+                    var propertyValues = ConvertToDictionary(value, GetPath(path, property.Name), debth++);
+                    foreach (var kvp in propertyValues)
+                    {
+                        result.Add(kvp.Key, kvp.Value);
+                    }
+                }
+                else if (property.PropertyType != typeof(string) && property.PropertyType.GetInterfaces().Any(x => x.IsGenericType && x.GetGenericTypeDefinition() == typeof(IEnumerable<>)))
+                {
+                    // Could handle IEnmerables here
```
Comment (virensareen) [inline VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs:43]: Comment?

File: VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs (hunk 2 above)
Comment (levibutler) [reply VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs:19]: 👍 will fix :)

File: VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs (hunk 3 above)
Comment (levibutler) [reply VideoWeb/VideoWeb.Common/Helpers/LoggingDataExtractor.cs:43]: 👍 will fix :)

**Review by virensareen — APPROVED** [review]

**Review by virensareen — APPROVED** [review]

**Pull request merged**